packages/actor-extract-links-extract-tree/lib/FilterNode.ts

```diff
@@ -0,0 +1,206 @@
+import { BindingsFactory } from '@comunica/bindings-factory';
+import { KeysInitQuery } from '@comunica/context-entries';
+import type { IActionContext } from '@comunica/types';
+import type * as RDF from 'rdf-js';
+import { Algebra, Factory as AlgebraFactory } from 'sparqlalgebrajs';
+import { isBooleanExpressionTreeRelationFilterSolvable } from './solver';
+import type { Variable } from './solverInterfaces';
+import type { ITreeRelation, ITreeNode } from './TreeMetadata';
+
+const AF = new AlgebraFactory();
+const BF = new BindingsFactory();
+
+/**
+ * A class to apply [SPAQL filters](https://www.w3.org/TR/sparql11-query/#evaluation)
+ * to the [TREE specification](https://treecg.github.io/specification/).
+ * It use [sparqlee](https://github.com/comunica/sparqlee) to evaluate the filter where
+ * the binding are remplace by the [value of TREE relation](https://treecg.github.io/specification/#traversing).
+ */
+export class FilterNode {
+  /**
+   * Return the filter expression if the TREE node has relations
+   * @param {ITreeNode} node - The current TREE node
+   * @param {IActionContext} context - The context
+   * @returns {Algebra.Expression | undefined} The filter expression or undefined if the TREE node has no relations
+   */
+  public getFilterExpressionIfTreeNodeHasConstraint(node: ITreeNode,
```

Comment:
This method is combining 2 things that probably shouldn't be combined here. (getting filter expr, and checking of there are node relationships)